Add screenshot service, devcontainer.

mattcburns opened the pull request:
## Description

The primary feature of this PR is to add screenshot support to the service that was earlier added to bmclib. In addition, I have added devcontainer support to make development easier on some platforms.

I have removed go-grpc validations from the protobuf definitions and grpc build process. This is due to incompatibility with building grpc libraries for languages other than Go. If there is some workaround (or I missed something), I am open to suggestions but otherwise it should be removed so other languages can use the client generation capabilities.

## How Has This Been Tested?
I have added a command line option under a new command subcategory called `diagnostic screenshot` to test capture of the screenshots via CLI. A user needs to set the `BMCIP` and `BMCPASS` environment variables and the command line tool will use these to connect to the target machine.


## How are existing users impacted? What migration steps/scripts do we need?

No major impact to end users - if they want to take advantage of the new options they will need to update any tooling around this application.


## Checklist:

I have:

- [X] updated the documentation and/or roadmap (if required)
- [X] added unit or e2e tests
- [X] provided instructions on how to upgrade

jacobweinstock commented:
Hey @mattcburns, thanks for the PR! I'm not convinced that adding this amount of Ruby code is a good idea for the repo. The existing Ruby code was only there as a very basic example of a client. This amount of code feels like it's a library and now something we'll need to maintain. I'm not familiar enough with Ruby to want to take this on. I don't believe the other maintainers are familiar with Ruby either but i'll let them chime in. cc @joelrebel , @chrisdoherty4

mattcburns commented:
> Hey @mattcburns, thanks for the PR! I'm not convinced that adding this amount of Ruby code is a good idea for the repo. The existing Ruby code was only there as a very basic example of a client. This amount of code feels like it's a library and now something we'll need to maintain. I'm not familiar enough with Ruby to want to take this on. I don't believe the other maintainers are familiar with Ruby either but i'll let them chime in. cc @joelrebel , @chrisdoherty4

I can just remove it, it's auto generated from the protobufs so we don't really need to keep it in there. Thanks for the feedback!

jacobweinstock commented on api/v1/bmc.proto:
```diff
@@ -19,7 +18,7 @@ service BMC {
 message NetworkSourceRequest {
     v1.Authn authn = 1;
     v1.Vendor vendor = 2;
-    NetworkSource network_source = 3 [(validator.field) = {is_in_enum : true}];
```
I'm not opposed to dropping these validations, I would just like to be sure that the validations are handled. I could be missing it, but i don't see them handled in any other way in the PR?

mattcburns replied:
I'm not sure how to support validations in the proto definitions without the library but with the library I can't generate client code for languages other than go. I'll do some more research to see what can be done for validations.
Do you happen to know what the original intent was behind adding the validations into the protobuf definitions?

mattcburns replied:
Well, I went back to this again on the ruby side and found the right way to build this while keeping the validations, I'll update the PR shortly to reflect this.

joelrebel reviewed (commented):
Nice work, left a few suggestions inline.

joelrebel commented on api/v1/bmc.proto:
```diff
@@ -19,7 +18,7 @@ service BMC {
 message NetworkSourceRequest {
     v1.Authn authn = 1;
     v1.Vendor vendor = 2;
-    NetworkSource network_source = 3 [(validator.field) = {is_in_enum : true}];
```
I see Jacob got to this before I did, will set this to resolved.

jacobweinstock commented on cmd/screenshot.go:
```diff
@@ -0,0 +1,60 @@
+						Host: &v1.Host{
+							Host: os.Getenv("BMCIP"),
+						},
+						Username: "root",
```
FYI, CLI flags are already available as env vars through viper as `PBNJ_`. reference: https://github.com/tinkerbell/pbnj/blob/629921c5f6e4282ef1a2281a8995e5801b0ddeaa/cmd/root.go#L62C1-L64C1

If you define cobra cli flags the envs will come along with the viper integration. here's the `server` command example: https://github.com/tinkerbell/pbnj/blob/629921c5f6e4282ef1a2281a8995e5801b0ddeaa/cmd/server.go#L109

mattcburns replied:
Ah good point, I'll add that to the cobra config then

jacobweinstock commented on cmd/diagnostic.go:
```diff
@@ -0,0 +1,16 @@
+)
+
+// diagnosticCmd represents the diagnostic command.
+var diagnosticCmd = &cobra.Command{
```
I'm not opposed to this but i don't understand the reason for this subcommand. `pbnj diagnostic screenshot` vs `pbnj screenshot`, right? curious what you're thinking/optimizing for here?

jacobweinstock replied:
oh maybe because the protobuf service is called `diagnostic`?

mattcburns replied:
The idea is that in the future there may be other diagnostic commands other than screenshot - but it may be too early to start doing that; I can simplify it down to just screenshot.

jacobweinstock commented on grpc/oob/diagnostic/screenshot.go:
```diff
@@ -0,0 +1,144 @@
+}
+
+// WithScreenshotRequest adds ScreenshotRequest to an Action struct.
+func WithScreenshotRequest(in *v1.ScreenshotRequest) Option {
```
This feels like it should be required always, no? if that is the case, you might consider making it explicit as a dependency in the `GetScreenshot` function.

mattcburns commented:
I addressed all the feedback - I didn't realize that a force push after I rebased would remove most of those inline comments, so I apologize for that, but I think they're all still part of the conversation view. Thank you!

joelrebel commented on grpc/oob/diagnostic/screenshot.go:
```diff
@@ -0,0 +1,138 @@
+
+	opts := []bmclib.Option{
+		bmclib.WithLogger(m.Log),
+		bmclib.WithPerProviderTimeout(45 * time.Second),
```
Theres a helper method here to retrieve the timeout set through env vars,

https://github.com/tinkerbell/pbnj/blob/629921c5f6e4282ef1a2281a8995e5801b0ddeaa/grpc/oob/common.go#L118

mattcburns commented:
Addressed the additional feedback - this should be ready for further review @jacobweinstock  @joelrebel 
Thanks!

jacobweinstock reviewed (approved):
Thanks for this @mattcburns !
